From 82a0d1f78604547b9388912e5d04d73e603dd303 Mon Sep 17 00:00:00 2001
From: Binh Nguyen <binh.nguyen.uw@rvc.renesas.com>
Date: Mon, 2 Jul 2018 18:05:39 +0700
Subject: [PATCH] quick: scenegraph: fix texture width alignment issue for
 PowerVR G6400

There was unclear crash issue when using PowerVR G6400 on iwg21m board

Signed-off-by: Binh Nguyen <binh.nguyen.uw@rvc.renesas.com>
---
 src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp b/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
index 0c470fc..c248e2d 100644
--- a/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
+++ b/src/quick/scenegraph/qsgdefaultdistancefieldglyphcache.cpp
@@ -100,6 +100,14 @@ void QSGDefaultDistanceFieldGlyphCache::requestGlyphs(const QSet<glyph_t> &glyph
         QRectF boundingRect = glyphData(glyphIndex).boundingRect;
         int glyphWidth = qCeil(boundingRect.width()) + distanceFieldRadius() * 2;
         int glyphHeight = qCeil(boundingRect.height()) + distanceFieldRadius() * 2;
+        /*
+        FIXME: There was unclear crash issue with PowerVR Rogue G6400 when width is odd.
+        Below add a workaround to prevent odd width when using PowerVR Rogue G6400.
+        */
+        if (!qstrcmp(reinterpret_cast<const char*>(m_funcs->glGetString(GL_RENDERER)),
+                                "PowerVR Rogue G6400") && glyphWidth%2 != 0)
+            glyphWidth+=1;
+
         QSize glyphSize(glyphWidth + padding * 2, glyphHeight + padding * 2);
         QRect alloc = m_areaAllocator->allocate(glyphSize);
 
-- 
2.7.4

